src/Microsoft.TestPlatform.CrossPlatEngine/TestSession/ProxyTestSessionManager.cs

```diff
@@ -382,6 +384,35 @@ private void DisposeProxies()
             _proxyMap.Clear();
         }
     }
+
+    private bool CheckRunSettingsAreCompatible(string requestRunSettings)
+    {
+        if (_runSettingsEnvironmentVariables.Count == 0)
+        {
+            _runSettingsEnvironmentVariables = InferRunSettingsHelper.GetEnvironmentVariables(
+                    _testSessionCriteria.RunSettings)
+                ?? _runSettingsEnvironmentVariables;
+        }
+
+        // Environment variable sets should be identical, otherwise it's not safe to reuse the
+        // already running testhosts.
+        var requestEnvironmentVariables = InferRunSettingsHelper.GetEnvironmentVariables(requestRunSettings);
+        if (requestEnvironmentVariables != null
+            && _runSettingsEnvironmentVariables != null
+            && (requestEnvironmentVariables.Count != _runSettingsEnvironmentVariables.Count
+                || requestEnvironmentVariables.Except(_runSettingsEnvironmentVariables).Any()))
+        {
+            return false;
+        }
+
+        // Data collection is not supported for test sessions yet.
+        if (XmlRunSettingsUtilities.IsDataCollectionEnabled(requestRunSettings))
```

Comment:
it would be good to figure out if this check can be relaxed and if so how

Reply:
The check can be relaxed and I was planning to do it, but it will require more work and that means adopting Test Sessions API would be blocked for longer. This check was only added to reflect the state of the art, i.e. we cannot have data collectors when starting the test session and because of that we cannot have them when matching the testhosts with test discovery/run requests. But again, this is only temporary.